src/v/cluster/archival/archiver_operations_impl.cc

```diff
@@ -299,6 +634,239 @@ class archiver_operations_impl : public archiver_operations_api {
+              .code = cloud_storage::upload_result::success,
+              .put_requests = 1,
+              .bytes_sent = ix_size,
+            };
+        } catch (...) {
+            vlog(
+              _rtclog.error,
+              "Failed to build an index, unexpected error: {}",
+              std::current_exception());
+            co_return aggregated_upload_result{
+              .code = cloud_storage::upload_result::failed};
+        }
+        vassert(false, "unreachable");
+    }
+
+    static bool segment_meta_matches_stats(
+      const cloud_storage::segment_meta& meta,
+      const cloud_storage::segment_record_stats& stats,
+      retry_chain_logger& ctxlog) {
+        // Validate segment content. The 'stats' is computed when
+        // the actual segment is scanned and represents the 'ground truth' about
+        // its content. The 'meta' is the expected segment metadata. We
+        // shouldn't replicate it if it doesn't match the 'stats'.
+        if (
+          meta.size_bytes != stats.size_bytes
+          || meta.base_offset != stats.base_rp_offset
+          || meta.committed_offset != stats.last_rp_offset
+          || static_cast<size_t>(meta.delta_offset_end - meta.delta_offset)
+               != stats.total_conf_records) {
+            vlog(
+              ctxlog.error,
+              "Metadata of the uploaded segment [size: {}, base: {}, last: {}, "
+              "begin delta: {}, end delta: {}] "
+              "doesn't match the segment [size: {}, base: {}, last: {}, total "
+              "config records: {}]",
+              meta.size_bytes,
+              meta.base_offset,
+              meta.committed_offset,
+              meta.delta_offset,
+              meta.delta_offset_end,
+              stats.size_bytes,
+              stats.base_rp_offset,
+              stats.last_rp_offset,
+              stats.total_conf_records);
+            return false;
+        }
+        return true;
+    }
+
+    /// Upload segment with index and tx-manifest
+    ///
+    /// The result is a combination of all upload results
+    /// plus the segment metadata generated during the index build
+    /// step.
+    ss::future<aggregated_upload_result> upload_candidate(
+      std::reference_wrapper<retry_chain_node> workflow_rtc,
+      ss::shared_ptr<cluster_partition_api> partition,
+      reconciled_upload_candidate_ptr upl) {
+        auto [segment_name, index_name, tx_name] = segment_index_object_names(
+          partition, upl);
+
+        cloud_storage::segment_record_stats stats{};
+
+        auto [index_stream, upload_stream] = clone_stream(
+          std::move(upl->payload), _readahead());
+
+        auto make_su_fut = upload_segment(
+          workflow_rtc, segment_name, upl, std::move(upload_stream));
+        auto make_ix_fut = upload_segment_index(
+          workflow_rtc, stats, upl, index_name, std::move(index_stream));
+
+        chunked_vector<ss::future<aggregated_upload_result>> uploads;
```

Comment:
nit: since the size is fixed, can this be `std::vector` reserved to 3?

Reply:
fixed